apps/frontend/tests/component_tests/Profile.cy.tsx

```diff
@@ -0,0 +1,215 @@
+        cy.log("Request Body", req.body);
+        // Validate the request body
+        expect(req.body).to.deep.equal(updatedProfile);
+        res.send({ statusCode: 400, body: { message: "Bad Request" } });
+      });
+    }).as("saveProfileRequest");
+
+    // Assert that the profile information remains as the updated values (failed to save does not revert the inputs)
+    cy.get("#name").should("have.value", updatedName);
+    cy.get("#bio").should("have.value", updatedBio);
+    cy.get("#username").should("have.value", testUsername);
+  });
+
+  it("cancels the changes and reverts to original profile information", () => {
+    // Stubbing network request for getting profile information
+    cy.intercept("GET", "/api/user/1", {
+      statusCode: 200,
+      body: {
+        username: testUsername,
+        name: testName,
+        password: testPassword,
+        bio: testBio,
+        profilePictureUrl: testImageURL,
+      },
+    }).as("profileRequest");
+
+    // Wait for the mock request to complete and check its status
+    cy.wait("@profileRequest").its("response.statusCode").should("eq", 200);
+
+    // Type into the input fields
+    cy.get("#name").clear().type(updatedName).should("have.value", updatedName);
+    cy.get("#bio").clear().type(updatedBio).should("have.value", updatedBio);
+
+    // Click the cancel button
+    cy.get("#cancel_button").click();
+
+    // Assert that the profile information is reverted correctly
+    cy.get("#name").should("have.value", testName);
+    cy.get("#bio").should("have.value", testBio);
+  });
+
+  it("disables buttons when no profile information has been updated", () => {
+    // Stubbing network request for getting profile information
+    cy.intercept("GET", "/api/user/1", {
+      statusCode: 200,
+      body: {
+        username: testUsername,
+        name: testName,
+        password: testPassword,
+        bio: testBio,
+        profilePictureUrl: testImageURL,
+      },
+    }).as("profileRequest");
+
+    // Wait for the mock request to complete and check its status
+    cy.wait("@profileRequest").its("response.statusCode").should("eq", 200);
+
+    // Check that save and cancel buttons are disabled
+    cy.get("#save_button").invoke("attr", "disabled").should("exist");
+    cy.get("#cancel_button").invoke("attr", "disabled").should("exist");
+  });
+
+  // Check error for handling invalid username format
+  it("displays an error message for invalid username format", () => {
+    // Type an invalid username
+    cy.get("#username")
+      .type("invalid-username2389012839082103820198190830981903")
+      .should(
+        "have.value",
+        "invalid-username2389012839082103820198190830981903"
+      );
+
+    // Click the save button
+    cy.get("#save_button").click();
+
+    // Check if the error message is displayed
+    cy.get("p").should(
+      "contain.text",
+      "Username must be between 1 and 20 characters and only contain letters or numbers."
+    );
+  });
```

Comment:
When running the tests, I seem to get some duplicate API requests triggered while typing the invalid username. Probably worth some investigation, but it could be something with my cypress as I can't see a reason that would trigger the get request.
 
![image](https://github.com/UVicMartletplace/martletplace/assets/54737724/f8fdd0a0-9a0b-4d78-b9fb-dfd1f971551a)